### ISSUE TITLE
host anti-affinity is not working 

### ISSUE THREAD

**Issue by kiranchavala** [apache/cloudstack#7202]
##### ISSUE TYPE

 * Bug report

##### COMPONENT NAME

~~~
Compute
~~~

##### CLOUDSTACK VERSION

~~~
4.18 nightly
4.17.2

~~~

##### OS / ENVIRONMENT

Centos 7 , 1 management , 1 pod , 1 cluster with 2 kvm hypervisor hosts

##### SUMMARY

host anti-affinity is not working 

##### STEPS TO REPRODUCE
~~~

1. Create a affinity group of type "host anti-affinity (Strict)".

2. Deploy 2 vm without any affinity group attached 

3. Check the vm's host ( they can be deployed on the same host)

Note: Make sure the vm's are deployed on 1 single host in the cluster

3. Stop the vm's

4. Change the affinity group  of the vms to  "host anti-affinity (Strict)"

Note: Make sure the affinity group is attached to the vm

5. Start the vm 

6. The Vm 's still remains on the same host in the cluster

Made sure to uncheck the option to start on Last host 
![start](https://user-images.githubusercontent.com/1401014/218453831-ed4070bf-707d-4d39-80fc-36477a7e74a2.png)



~~~

##### EXPECTED RESULTS

~~~
The vm's should go to different host in the cluster
~~~

##### ACTUAL RESULTS

~~~
The vm's  are not going  to different hosts  in the cluster
~~~

**Comment by kiranchavala** [comment]
@DaanHoogland The issue is also observed on 4.17.2 release.

**Comment by kiranchavala** [comment]
@DaanHoogland 

The issue is not observed if we deploy directly instances with an affinity group attached. (First attempt)

The issue is observed when the affinity group is attached or changed to the instance on subsequent attempts

**Comment by weizhouapache** [comment]
@kiranchavala 
Have you tried with considerlasthost = false ?

![image](https://user-images.githubusercontent.com/57355700/218667177-55d8cea6-f984-4ef5-a4c4-a481fe05475a.png)

**Comment by kiranchavala** [comment]
@weizhouapache I didn't try with consider Last host ( it was disabled)

**Comment by weizhouapache** [comment]
@kiranchavala I will try to reproduce the issue

**Comment by weizhouapache** [comment]
@kiranchavala 
I cannot reproduce the issue. see steps below
- create host anti-affinity group
- create vm1 on kvm1, with host anti-affinity group
- create vm2 on kvm1, without host anti-affinity group
- stop vm2
- start vm2 with considerlasthost=false (by cmk)
- vm2 is started on kvm2

**Comment by kiranchavala** [comment]
@weizhouapache 

Could you please follow these steps 

~~~

1. Create a affinity group of type "host anti-affinity (Strict)".

2. Deploy 2 vm's ok  without any affinity group attached 

3. Check the vm's host ( they can be deployed on the same host)

Note: Make sure the vm's are deployed on 1 single host in the cluster

3. Stop the vm's

4. Change the affinity group  of the vms to  "host anti-affinity (Strict)"

Note: Make sure the affinity group is attached to the vm

5. Start the vm (Made sure to uncheck the option to start on Last host )

6. The Vm 's still remain on the same host in the cluster in spite of changing the affinity group

~~~

**Comment by weizhouapache** [comment]
@kiranchavala 
I still cannot reproduce the issue you mentioned.

but I found that if hostid is specified, the affinity groups  seems not be checked

**Comment by weizhouapache** [comment]
@kiranchavala 
I am able to reproduce the issue
I will have a look

**Comment by weizhouapache** [comment]
@kiranchavala 
this only happens when start both VMs in parallel in my testing.
If I start the vms one by one, it works fine.

is that same in your testing ?

**Comment by weizhouapache** [comment]
moved to 4.18.2.0

**Comment by DaanHoogland** [comment]
@kiranchavala @weizhouapache I think this is a logical consequence of the functionality. If you add an affinity group after creating, the system must have the chance to find a proper location for the VM, it does so with two VM in parallel and for each the location is proper at the moment of starting to only be in conflict with a parallel start. I will do more testing (with initial starting of a present anti affinity group) to see if more strange things occur, but I think this one is due to the operator action and to be expected.

**Comment by rohityadavcloud** [comment]
This is still an issue, @kiranchavala ?

**Comment by DaanHoogland** [comment]
@kishankavala a solution has been merged in 4.19 branch in #9201 . Can you verify it?

**Comment by kiranchavala** [comment]
The issue is fixed, , followed these steps and its working fine

```

1. Create a affinity group of type "host anti-affinity (Strict)".

2. Deploy 2 vm's ok  without any affinity group attached 

3. Check the vm's host ( they can be deployed on the same host)

Note: Make sure the vm's are deployed on 1 single host in the cluster

3. Stop the vm's

4. Change the affinity group  of the vms to  "host anti-affinity (Strict)"

Note: Make sure the affinity group is attached to the vm

5. Start the vm (Made sure to uncheck the option to start on Last host )

6. The vm's got deployed  on different hosts 
```
